pkg/controller/spec_processor.go

```diff
@@ -112,81 +130,59 @@ func (sp *SpecProcessor) ProcessValue(value interface{}, path ...string) (interf
 }
 
 func (sp *SpecProcessor) ProcessString(value string, path ...string) (interface{}, error) {
-	if !reference.MatchString(value) {
+	match := reference.FindStringSubmatch(value)
+	if match == nil {
 		return value, nil
 	}
 
-	result, err := sp.processMatch(value[2 : len(value)-2])
-	if err != nil {
-		return nil, errors.Wrapf(err, "invalid reference at \"%s\"", strings.Join(path, "."))
+	// This means we have multiple '!' at the start, so we're escaping.
+	if len(match[1]) > 1 {
```

Comment:
Agree re:tests. I would like to see examples of such escaping format.

Reply:
Sorry, I thought I added tests.
I don't see a problem with this method of escaping though - it should just work, aside from the programmatic construction case... my preference here, rather than a flag, would be to look for references IFF there are references (i.e. if variables is non-empty).

Reply:
(yes, it's a little magical, but it keeps the base case clean)

Reply:
On further thought, I think the current behaviour is the correct one. Programmatic users should just have to deal with escaping (it's relatively straightforward, as are the error cases... we also have no cases currently).

Reply:
I will add the functions to escape strings/objects in a separate PR?